src/serial.rs

```diff
@@ -0,0 +1,474 @@
+            DATA => {
+                // Here we emulate the reset method for when RDA interrupt
+                // was raised (i.e. read the receive buffer and clear the
+                // interrupt identification register and RDA bit when no
+                // more data is available).
+                self.del_interrupt(IIR_RDA_BIT);
+                if self.in_buffer.len() <= 1 {
+                    self.clear_lsr_rda_bit();
+                }
+                self.in_buffer.pop_front().unwrap_or_default()
+            }
+            IER => self.interrupt_enable,
+            IIR => {
+                // This is an interesting trick to let the driver be able
+                // to read many bytes without complaining (by simulating
+                // the FIFO capability).
+                let iir = self.interrupt_identification | IIR_FIFO_BITS;
+                self.iir_reset();
+                iir
+            }
+            LCR => self.line_control,
+            MCR => self.modem_control,
+            LSR => self.line_status,
+            MSR => {
+                if self.is_in_loop_mode() {
+                    // All of these things are done because of the way the loopback test works
+                    // for COM4. More details in commit 207b1fd from crosvm.
+                    let mut msr =
+                        self.modem_status & !(MSR_DSR_BIT | MSR_CTS_BIT | MSR_RI_BIT | MSR_DCD_BIT);
+                    if (self.modem_control & MCR_DTR_BIT) != 0 {
+                        msr |= MSR_DSR_BIT;
+                    }
+                    if (self.modem_control & MCR_RTS_BIT) != 0 {
+                        msr |= MSR_CTS_BIT;
+                    }
+                    if (self.modem_control & MCR_OUT1_BIT) != 0 {
+                        msr |= MSR_RI_BIT;
+                    }
+                    if (self.modem_control & MCR_OUT2_BIT) != 0 {
+                        msr |= MSR_DCD_BIT;
+                    }
+                    msr
+                } else {
+                    self.modem_status
+                }
+            }
+            SCR => self.scratch,
+            _ => 0,
+        }
+    }
+
+    /// Helps in sending more bytes to the guest in one shot, by storing
+    /// `input` bytes in UART buffer and letting the driver know there is
+    /// some pending data to be read by setting RDA bit and its corresponding
+    /// interrupt when not already triggered.
+    ///
+    /// # Arguments
+    /// * `input` - The data to be sent to the guest.
+    pub fn queue_raw_bytes(&mut self, input: &[u8]) -> Result<()> {
+        if !self.is_in_loop_mode() {
+            self.in_buffer.extend(input);
+            self.set_lsr_rda_bit();
+            self.received_data_interrupt()?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::io::sink;
+    use std::sync::{Arc, Mutex};
+
+    #[derive(Clone)]
+    struct SharedBuffer {
```

Comment:
Can we avoid defining the `SharedBuffer` struct, and simply use a `Vec<u8>` instead?

Reply:
yep :100: